### ISSUE TITLE
[$1000] [HOLD for payment 2024-07-24] [Feature] Improve 'location' permission flow with explainer and denied handling modals

### ISSUE THREAD

**Issue by Julesssss** [Expensify/App#37736]
### Problem

Currently, we do not handle the case where users deny the location permission. This leads to a few problems on our mobile apps:

- After requesting money, users are prompted for the location permission _every single time_ which is confusing, invasive, and annoying
- If the user denies this permission, we are restricted from asking them again
    - iOS: We cannot ask again after the first denial
    - Android: We cannot ask again after the 2nd denial
- When restricted ^, we cannot deep link users to the permission settings page (on Android), we can only explain the actions they must take
- The current flow looks like a [bug](https://github.com/Expensify/App/issues/36882) as we are asking for the location permission after the user has left the Request Money flow
- Android users who Scan a receipt, turn on location when requested, then accept the location permission remain stuck at the final step in the Request flow

**GPS disabled, location permission disabled**

iOS | Android
-- | --
![E7A805B3-AF24-4C42-8954-48A25C4E8AFE](https://github.com/Expensify/App/assets/10736861/503013a4-89ed-48f8-b720-c36a48deb7c4) | ![Screenshot_20240305-114302](https://github.com/Expensify/App/assets/10736861/10d8af3d-e493-4873-935f-c5cccfa54dd5)

**GPS enabled, location permission disabled**

iOS | Android
-- | --
![9CECE423-2AD8-4A9F-9505-629741B51406](https://github.com/Expensify/App/assets/10736861/8aaeeb51-6d37-45d2-9c13-1ea667e0559d) | ![Screenshot_20240305-113953](https://github.com/Expensify/App/assets/10736861/b6bc08c3-2ef8-4f0c-aea9-2291c976ffc3)

**Full location flow (Android)**

https://github.com/Expensify/App/assets/10736861/1bbec545-1702-4041-b015-955e913dcc1b

### Solution

The typical solution for mobile apps is to implement the following for permissions:
1. We show a custom page/modal explaining why we are about to ask them to enable a permission and offering a choice(something like `Okay` or `Cancel`)
2. If they tap `Okay`, we show them the native OS permission prompt
3. If they tap 'Cancel', we do not show the native OS permission prompt
4. If we are denied asking the user, show a custom modal explaining how they can manually allow the permission from the OS App settings

_This better explains to users why we need the prompt, and prevents the need for the complicated 'Please navigate to App Settings > Permissions, and enable XYZ' instructions._

___ 

We have already implemented the above flow for the Camera permission. Let's build a similar flow for the location prompt. Here is the Camera flow:

x | iOS | Android
-- | -- | --
Explainer | ![87A5E029-E433-43EA-A576-BD65FEFE159B](https://github.com/Expensify/App/assets/10736861/463a5852-fa3b-4173-bb42-2474f6650b1a) | ![Screenshot_20240305-112003](https://github.com/Expensify/App/assets/10736861/921eca7e-626d-4946-a8e2-0acb39871410)
Native OS prompt | ![E4ED35A0-91E2-4409-9359-6DC526D31727](https://github.com/Expensify/App/assets/10736861/fb6e5c7b-023c-478f-a309-f17bb00a9d7e) | ![Screenshot_20240305-113921](https://github.com/Expensify/App/assets/10736861/26425085-f90e-4f34-a38b-c581751bd71e)
Denied | ![CC59EB0E-61E2-4596-A942-28A57242A9D9](https://github.com/Expensify/App/assets/10736861/19891cfa-ba46-4a2a-b548-5b066f49beab) | ![Screenshot_20240305-112013](https://github.com/Expensify/App/assets/10736861/698b0310-60fb-49af-9bde-edab238b34ce)

### Proposed changes: 

_Pending @Expensify/design review_

**Location permission flow**
- When 'Request' button is pressed at the end of the Scan Receipt flow, check for location permission:
    - **If granted**, continue as normal
    - **If not granted**, show a modal that explains what/why we about to ask for the users location, and give two options: 'Okay, Cancel'
        - **Okay:** Show the native permission prompt, continue as normal
            - **NOTE:** The OS will show an additional modal here if GPS is not enabled 
        - **Cancel:** Continue as normal, without users locations
    - **If we are denied from asking**, show a 'Denied' modal that explains how users can re-enable the permission manually 

```
- GPS disabled, permission disabled, GPS prompt closed
- GPS disabled, permission disabled, modal closed
- GPS enabled, permission enabled, permission prompt closed
- GPS enabled, permission enabled, permission prompt rejected
- GPS enabled, permission enabled, permission prompt approved
- Permission prompt rejected (once iOS, twice Android), disabled modal accepted
- Permission prompt rejected (once iOS, twice Android), disabled modal rejected
```

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ae8e8fb606975e83</li>
        <li>Upwork Job ID: 1764983541349486592</li>
        <li>Last Price Increase: 2024-07-19</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @zanyrenney</details>

**Comment by Julesssss** [comment]
Hi @Expensify/design team. Please share your thoughts and suggestions!
- One difference between the camera and location flow is that the camera explainer is displayed nicely within the in-page camera component, location doesn't have a similar in-page component so I assume we'd need to use a Modal instead?
- Asking for the location permission every time the user scans a receipt can get quite annoying. What are your thoughts here? Should we stick with this for now, or explore alternatives like a timed delay?

**Comment by twisterdotcom** [comment]
>Asking for the location permission every time the user scans a receipt can get quite annoying. What are your thoughts here? Should we stick with this for now, or explore alternatives like a timed delay?

I agree, we don't do this in oldDot. I think we should ask once and if they choose "once" or "don't allow", we shouldn't ask again for SmartScan. (We can ask again for distance or anything else that may trigger it).

**Comment by dannymcclain** [comment]
Definitely agree with the planned course of action here—very standard 👍

>One difference between the camera and location flow is that the camera explainer is displayed nicely within the in-page camera component, location doesn't have a similar in-page component so I assume we'd need to use a Modal instead?

Yeah I'm thinking a modal will be best too. Definitely open to other ideas from the design team or anyone else.
When/where would that modal pop up?

Do we have any "bottom docked" modal patterns in NewDot that we could use for this? Might be nicer than a full page takeover (on desktop I imagine it's still just going to be a small, centered modal).

>I agree, we don't do this in oldDot. I think we should ask once and if they choose "once" or "don't allow", we shouldn't ask again for SmartScan. (We can ask again for distance or anything else that may trigger it).

Yeah maybe we should ask once per type of flow? (scan, distance, etc.)

**Comment by shawnborton** [comment]
> Do we have any "bottom docked" modal patterns in NewDot that we could use for this?

Yup! It might be bottom-docked on mobile and then a centered modal on desktop for that pattern.

I agree with the general sentiment though - ask each time for location (assuming the user is tapping "My location"), and only ask once for SmartScan. We might consider re-asking after some time but we shouldn't ask every single time.

**Comment by Julesssss** [comment]
Excellent, thanks everyone. So the next step will be for the flow to be designed. @dannymcclain are you happy to do this?

It's not super high priority but it would be great to get this implemented (by a contributor) for the collect release. I believe we need the following: 

- Component UI for explainer location permission prompt
- Component UI for declined location permission prompt
- (optional) update to existing modal for declined camera permission
- Copy and assets for the above

**Comment by rushatgabhane** [comment]
@Julesssss maybe you could add me back as C+ if this goes external

**Comment by trjExpensify** [comment]
I agree asking for location permission every time you scan a receipt is
annoying (and doesn’t it harm us if they keep saying no?), but equally it
does greatly improve the results - so a one and done worries me a little.

Bit of a spitball, but I wonder if we should explore doing something with
the system message on SmartScan failures? So when it fails and you haven’t
granted us location permissions, we surface that somehow in the expense
thread as a protip to improve results?

**Comment by dannymcclain** [comment]
Sounds good to me! I'll share some stuff here once I've had a chance to work on it.

**Comment by dannymcclain** [comment]
How about something like this? cc @Expensify/design 

![image](https://github.com/Expensify/App/assets/7516624/72c63466-a925-4d31-87e5-0f6709f46332)

![image](https://github.com/Expensify/App/assets/7516624/77d96047-1f7d-4c5a-8634-b1ef186334ee)

> [!NOTE]
>
> Definitely need help with the copy.
> For desktop, is it ok to throw a modal on top of another modal (RHP)?

**Comment by shawnborton** [comment]
Nice! I like 1 and 3 here. The simplicity of 1 is quite nice, though you lose the cool earth. What are you feeling?

**Comment by dannymcclain** [comment]
I think I like 3 the best. 1 is nice and simple but doesn't say "location" enough to me.

**Comment by shawnborton** [comment]
Cool cool, makes sense to me.

**Comment by dannymcclain** [comment]
@Julesssss (and everyone else) does this look right to you in terms of the flow for the "happy path"?

![image](https://github.com/Expensify/App/assets/7516624/29f0e53f-b458-4bfa-a5dc-a3740b173ffc)

A couple questions:
- I'm assuming we show the "soft ask" bottom modal after a user hits the "Request" button—is that correct?
- If yes to the above, once the user grants us permission for their location, do we charge ahead with the request or will the user need to hit the "Request" button again?

**Comment by Julesssss** [comment]
> I'm assuming we show the "soft ask" bottom modal after a user hits the "Request" button—is that correct?

Yep, correct.

> If yes to the above, once the user grants us permission for their location, do we charge ahead with the request or will the user need to hit the "Request" button again?

Yep, correct. We'll move forward regardless of their response and fallback to ip location if they don't grant us. 

One small thing: For the 'Allow location explainer', can we specifically mention that this allows us to more accurately detect their expense currency?

**Comment by dannymcclain** [comment]
>One small thing: For the 'Allow location explainer', can we specifically mention that this allows us to more accurately detect their expense currency?

1000%. I just threw some made up silliness in there. We're going to need actual copy for sure. Maybe something like `We'll use your location to help accurately determine your default currency and timezone. You can edit access in your device's settings anytime.`

**Comment by dannymcclain** [comment]
Ok @Julesssss, looking at [your list](https://github.com/Expensify/App/issues/37736#issuecomment-1980489643), do these screens make sense?

![image](https://github.com/Expensify/App/assets/7516624/e20dcd4e-c680-4568-8c59-d2328f4a5f73)

![image](https://github.com/Expensify/App/assets/7516624/837f2141-1461-48e2-a3a3-aea9396600d2)

![image](https://github.com/Expensify/App/assets/7516624/66b2ff6d-5622-400c-ba28-6546814427cc)

I'm not sure I 100% understand the declined permission prompts—so please correct me if I've gotten something wrong!

**Comment by Julesssss** [comment]
That all looks great to me. I love the camera bird 😄. One small note is that the method for updating permissions is slightly different on iOS & Android, so maybe we would want to use a different string as the process is a bit tricky on Android:

- **iOS:** You land at a page with the location/camera toggle 
- **Android:** You land at the app settings page, then press 'Permissions' from a list, then must press 'Location', then press 'Allow only while using the app' or 'Ask every time'

1 | 2 | 3
-- | -- | --
![Screenshot_20240311-105254](https://github.com/Expensify/App/assets/10736861/836f0c2d-93cb-495b-a630-69053834a9a3) | ![Screenshot_20240311-105258](https://github.com/Expensify/App/assets/10736861/71ba188b-d51c-4a67-9a02-aded93373f82) | ![Screenshot_20240311-105553](https://github.com/Expensify/App/assets/10736861/f93c1e1b-01a4-49a2-8b2f-15829aea249d)

**Comment by dannymcclain** [comment]
@Julesssss yeah I think we should get some help with the copy in that case. From a design standpoint, what else do we need for this?

**Comment by Julesssss** [comment]
That looks good tbh, other than copy we'll just need the icon assets iin Figma and I'll export them.

Theres one other. case, but I'm mot sure we need to care about it: 
- User has given location permission
- User has GPS location OFF
- user scans receipt
- They are shown a native 'do you want to enable GPS' prompt with no context and they might have forgottern the original reason
- We could show the explainer modal with minor changes to the copy, or we could just leave it for now

**Comment by dannymcclain** [comment]
So in that situation they basically just have location services disabled globally? If that's the case I think we should just leave it alone for now.

**Comment by Julesssss** [comment]
We're discussing some final requirements [here](https://expensify.slack.com/archives/C036QM0SLJK/p1710428325101649?thread_ts=1710370709.576549&cid=C036QM0SLJK), then we'll be ready for implementation.

**Comment by Julesssss** [comment]
- We are ready to implement after the marketing team copy review
- We have a [separate issue ](https://github.com/Expensify/Expensify/issues/378927)for implementing the delayed prompt logic
- @dannymcclain could you please share assets, or provide a link we can use to export them?

We're pretty close to being able to implement this, FYI @rushatgabhane

**Comment by dannymcclain** [comment]
Everything should be available in [this Figma file](https://www.figma.com/file/WBDoJXmJyPpIYKpkJFb6cv/Location-Permission?type=design&node-id=35-3113&mode=design&t=w71QCjYPAXwWmADF-11).

I went ahead and exported the two illustration assets: [Location Permission Illustrations.zip](https://github.com/Expensify/App/files/14605671/Location.Permission.Illustrations.zip)

**Comment by rushatgabhane** [comment]
Okayy thank you!

**Comment by danielrvidal** [comment]
Here is what I would suggest:

Screen 1 - just removed “help” as we can be matter of fact.
> Allow location access
> We’ll use your location to accurately determine your default currency and timezone. You can edit access in your device’s settings anytime.

Screen 2 - called out we need to be in their settings. 

> Enable location in settings
> Allowing location access is required to help accurately determine your default currency and timezone. Tap Settings to update permissions. 

Screen 3 looks good to me.

**Comment by dannymcclain** [comment]
I updated the mocks in Figma to include Daniel's latest copy.

**Comment by Julesssss** [comment]
Excellent, thanks everyone. @rushatgabhane let us know if you have any questions, or feel free to DM me

**Comment by Julesssss** [comment]
FYI @rushatgabhane is now unavailable so @allroundexperts is going to implement the feature.

**Comment by allroundexperts** [comment]
Hi @Julesssss!
I wasn't able to work on it the previous weekend and had to leave it half way because of a household chore that came up unexpectedly. I think this would be ready by this weeks end.

**Comment by Julesssss** [comment]
Okay, sounds good. Thanks

**Comment by allroundexperts** [comment]
Update: The whole flow is ready. Just need to double check the icon sizes and margins but I don't have Figma access. I requested the access today. @dannymcclain can you please grant me those?

**Comment by Julesssss** [comment]
Hey @allroundexperts I can help you with that. 

@dannymcclain is it expected the sizes are slightly different? Should we add a frame to make them 100% identical? The RN View will probably be used for both so having the same size would be helpful. 

<img width="247" alt="Screenshot 2024-04-02 at 12 31 28" src="https://github.com/Expensify/App/assets/10736861/45b520d7-62a1-432a-8312-e33d610058ec">

**Comment by dannymcclain** [comment]
Good call @Julesssss. I've re-exported them at a standard size (160x140px)

- [Spy Pigeon with Camera Head.svg.zip](https://github.com/Expensify/App/files/14837054/Spy.Pigeon.with.Camera.Head.svg.zip)
- [Globe with Receipt and Location Marker.svg.zip](https://github.com/Expensify/App/files/14837056/Globe.with.Receipt.and.Location.Marker.svg.zip)

@allroundexperts I've also granted you access to that file, but it might not be super helpful anymore haha. Sorry about that!

**Comment by Julesssss** [comment]
I want the spy pigeon as my display profile, it perfectly represents my life in London -- Pidgeons and surveillance

**Comment by Julesssss** [comment]
How it going @allroundexperts? It would be great to get this into review as I'll be OOO starting Friday.

**Comment by allroundexperts** [comment]
Sorry, I was a little occupied on the weekend. I'll have it ready today.

**Comment by allroundexperts** [comment]
Created a draft PR https://github.com/Expensify/App/pull/39875

@Julesssss Feel free to begin the review to speed things up.

**Comment by Julesssss** [comment]
PR in review

**Comment by Julesssss** [comment]
Hey @Expensify/design, the PR for location permission flow improvements can be tested with [these links](https://github.com/Expensify/App/pull/39875#issuecomment-2084661801). _I'm using the design tag as you have all previously commented on this issue_

**Comment by dubielzyk-expensify** [comment]
Left comments. 2 issues on my end. One being a crash.

**Comment by Julesssss** [comment]
Hi @allroundexperts, are you able to provide an update? I'm pretty sure now that the crash is due to these changes -- even though we don't have the full reproduction steps.

**Comment by allroundexperts** [comment]
Hi @Julesssss!
I'll finalise this today.

**Comment by Julesssss** [comment]
PR crash was fixed. I noticed that the camera pidgeon 'GO TO SETTINGS TO ENABLE' modal isn't being disabled.

**Comment by garrettmknight** [comment]
Joining as bugzero.

**Comment by Julesssss** [comment]
> Joining as bugzero.

Thanks, this is an ideal assignment. I'm trying to re-work the location handling flow for the snack expenses that I submit to you.

**Comment by Julesssss** [comment]
Hey @allroundexperts, it would be great to get an update here. Ideally this would be focused on instead of new contributor issues if possible.

**Comment by allroundexperts** [comment]
Hi!
I came back from vacation today. This is on top of my list for today.

**Comment by Julesssss** [comment]
Hey @NickTooker, this issue is introducing a new flow where we warn the user why we are about to ask for their location permission (to make the receipt currency more successful), [here](https://github.com/Expensify/App/blob/9fe0582b4e6b0af273bfb6be6c90cf614d865651/src/languages/en.ts#L606-L609) are the 4 new sentences, plus [this](https://github.com/Expensify/App/blob/9fe0582b4e6b0af273bfb6be6c90cf614d865651/src/languages/en.ts#L123) small button copy ('Not now').

Here are some screenshots that should give you a better idea of the flow. let us know if you have any questions.

<img width="391" alt="Screenshot 2024-06-27 at 14 25 38" src="https://github.com/Expensify/App/assets/10736861/f6046c97-0366-4ea3-aceb-dfa1461ed14f">
<img width="390" alt="Screenshot 2024-06-27 at 14 25 30" src="https://github.com/Expensify/App/assets/10736861/5e136e7e-8768-4e8e-bf5e-cbda6e2d7dd2">

**Comment by NickTooker** [comment]
I prefer the copy and tone of the right screenshot. I think it fits our overall tone better than the left.

**Comment by NickTooker** [comment]
I think the copy and flow are spot on! 👍 from me to move forward with the above.

**Comment by Julesssss** [comment]
Thanks Nick

**Comment by Julesssss** [comment]
I requested Spanish translations review [here](https://expensify.slack.com/archives/C21FRDWCV/p1719572628293989)

**Comment by Julesssss** [comment]
I bumped the PR and translations requests today

**Comment by Julesssss** [comment]
Still awaiting translations. I bumped the request again.

**Comment by Julesssss** [comment]
We have a small [change request](https://github.com/Expensify/App/pull/39875#discussion_r1668774184) for the Spanish translations.

**Comment by Julesssss** [comment]
Bumping the PR for changes/review again today.

**Comment by Julesssss** [comment]
PR was merged today 🎉

**Comment by Julesssss** [comment]
PR hit staging and I tested with a real expense today. It's working well.

**Comment by Julesssss** [comment]
I see [this issue](https://github.com/Expensify/App/issues/26499) for the camera permission flow. I think in that case it was already partially implemented. But let me know if you can locate the issue or remember it differently.

**Comment by Julesssss** [comment]
A minor [issue](https://github.com/Expensify/App/issues/45431) was found in the flow during testing. Lets get that resolved before figuring out payment

**Comment by allroundexperts** [comment]
> A minor [issue](https://github.com/Expensify/App/issues/45431) was found in the flow during testing. Lets get that resolved before figuring out payment

@Julesssss I think we discussed this in the PR and decided to leave it as it is.

https://github.com/Expensify/App/pull/39875#issuecomment-2165523138

Do you have the updated requirements on how to handle this?

**Comment by Julesssss** [comment]
Oh thanks for reminding me. Once QA have verified the reported bug (the video seems to not match the bug) I will mark closed as expected.

**Comment by garrettmknight** [comment]
@shubham1206agra offer out to you.
Upwork Job: https://www.upwork.com/ab/applicants/1813618903142113868/hires?filter=shub&page=1

**Comment by Julesssss** [comment]
We also need to discuss with @allroundexperts the price for this feature implementation. The scope was bigger than usual, and although the complication delays shouldn't be considered we agreed beforehand that this would be treated differently.

**Comment by allroundexperts** [comment]
> We also need to discuss with @allroundexperts the price for this feature implementation. The scope was bigger than usual, and although the complication delays shouldn't be considered we agreed beforehand that this would be treated differently.

@Julesssss I'm happy with anything above $500 😄

**Comment by Julesssss** [comment]
I'd say $1000 seems fair considering the amount of work necessary here. We also started this a few months back, when the default bounty was higher.

Leaving this for a couple of days to see if anyone disagrees.

**Comment by garrettmknight** [comment]
@zanyrenney I've created the QA - can you just make sure these two get paid while I'm OOO? Thanks!

**Comment by garrettmknight** [comment]
Paid! Just waiting on @allroundexperts to request now

**Comment by allroundexperts** [comment]
I've noted it down @garrettmknight. Feel free to close!

**Comment by Julesssss** [comment]
> I've noted it down @garrettmknight. Feel free to close!

@allroundexperts I was about to close this issue, but I'm not sure what you meant by 'I've noted it down'? Can I assume you have received payment?

**Comment by allroundexperts** [comment]
I haven't received it yet. But I've added it to my tracker so even if its closed, I'd be able to still request the payment later on. I just do this to batch several payments together.

**Comment by Julesssss** [comment]
Oh great, thanks for confirming.

**Comment by JmillsExpensify** [comment]
$1,000 approved for @allroundexperts
